src/utils/utils.datasource.ts

```diff
@@ -13,3 +18,89 @@ export function isPrometheusDataSource(input: unknown): input is PrometheusDatas
     typeof input.uid === 'string'
   );
 }
+
+type DataSourceType = 'prometheus' | 'loki';
+
+/**
+ * Fetches and caches healthy data sources
+ */
+export class DataSourceFetcher {
+  private readonly pendingRequests = new Map<DataSourceType, Promise<DataSource[]>>();
+  private readonly cache = new Map<DataSourceType, DataSource[]>();
+
+  /**
+   * Retrieves healthy data sources of the specified type
+   * Results are cached indefinitely until the `DataSourceFetcher` is destroyed
+   *
+   * @param type - The type of data source to retrieve ('prometheus' or loki)
+   * @returns Array of healthy data sources
+   */
+  public async getHealthyDataSources(type: DataSourceType): Promise<DataSource[]> {
+    // Check if we have cached results
+    const cachedDataSources = this.cache.get(type);
+    if (cachedDataSources?.length) {
+      return cachedDataSources;
+    }
+
+    // If there's already a pending request for this type, wait for it
+    let pendingRequest = this.pendingRequests.get(type);
+    if (!pendingRequest) {
+      pendingRequest = this.fetchHealthyDataSources(type).finally(() => {
+        // Clean up the pending request after it completes
+        this.pendingRequests.delete(type);
+      });
+      this.pendingRequests.set(type, pendingRequest);
+    }
+
+    // Wait for the request to complete and update cache
+    const dataSources = await pendingRequest;
+    this.cache.set(type, dataSources);
+
+    return dataSources;
+  }
+
+  /**
+   * Fetches healthy data sources of the specified type
+   */
+  private async fetchHealthyDataSources(type: DataSourceType): Promise<DataSource[]> {
+    const allDataSourcesOfType = getDataSourceSrv().getList({
+      logs: true,
+      type,
+      filter: (ds) => ds.uid !== 'grafana',
+    });
+
+    const healthyDataSources: DataSource[] = [];
+    const unhealthyDataSources: DataSource[] = [];
+
+    await Promise.all(
+      allDataSourcesOfType.map(async (ds) => {
+        try {
+          const health = await getBackendSrv().get(`/api/datasources/uid/${ds.uid}/health`, undefined, undefined, {
+            showSuccessAlert: false,
+            showErrorAlert: false,
+          });
+
+          if (health?.status === 'OK') {
+            healthyDataSources.push(ds);
+          } else {
+            unhealthyDataSources.push(ds);
+          }
+          // NOTE: This appears to be a false positive from sonarjs
+          // eslint-disable-next-line sonarjs/no-ignored-exceptions
+        } catch (error) {
+          unhealthyDataSources.push(ds);
+        }
```

Comment:
Actually SonarJS is right. Have you tried:

```suggestion
        } catch {
          unhealthyDataSources.push(ds);
        }
```
?

Reply:
it works! https://github.com/grafana/metrics-drilldown/pull/384/commits/dbd65840f2ebfbb272a43c5253ea1ef7740612dd